lib/grpnames.gi

```diff
@@ -670,56 +671,137 @@ InstallGlobalFunction(DirectFactorsOfGroupKN,
+      and ForAll( Set(arg[2]), N ->
+                              IsSubgroup(arg[1], N) and IsNormal(arg[1], N))
+      and arg[3] in ["all", "any", "str"] then
+    G := arg[1];
+    Ns := ShallowCopy(arg[2]);
+    method := arg[3];
+    fullNs := false;
+  else
+    Error("usage: SemidirectDecompositionsOfFiniteGroup(<G> [, <Ns>] [, <mthd>])");
+  fi;
+
+  if HasSemidirectDecompositions(G) then
+    NHs := [ ];
+    for NH in SemidirectDecompositions(G) do
+      N := NH[1];
+      H := NH[2];
+      if method in [ "any", "str" ] and not IsTrivial(N)
+        and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      elif method="all" and
+        ( not IsBound(Ns) or N in Ns ) then
+        AddSet(NHs, [ N, H ]);
+      fi;
+    od;
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return NHs;
+    fi;
+  fi;
+
+  if method in [ "any", "str" ] then
+    if HasSemidirectProductInfo(G) then
+      N := Image(Embedding(G, 2));
+      H := Image(Embedding(G, 1));
+      if not IsTrivial(N) and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      fi;
+    fi;
+    N := NormalHallSubgroupsFromSylows(G, "any");
+    if N <> fail then
+      # by the Schur-Zassenhaus theorem there must exist a complement
+      if method = "any" then
+        H := ComplementClassesRepresentatives(G, N)[1];
+        return [ N, H ];
+      # only the isomorphism type of the complement is interesting
+      elif method = "str" then
+        Assert(1, Length( ComplementClassesRepresentatives(G, N) ) > 0);
+        return [ N, G/N ];
+      fi;
+    fi;
+  fi;
+
+  # simple groups have no nontrivial normal subgroups
+  if IsSimpleGroup(G) then
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return [ [TrivialSubgroup(G), G], [G, TrivialSubgroup(G)] ];
+    fi;
+  fi;
+
+  if not IsBound(Ns) then
+    Ns := ShallowCopy(NormalSubgroups(G));
+  fi;
+# does not seem to make things faster
+#  if method in [ "any", "str" ] then
+#    sizes := List(Ns, Size);
+#    SortParallel(sizes, Ns);
+#    Ns := Reversed(Ns);
+#  fi;
+
+  NHs := [ ];
+  for N in Ns do
+    H := ComplementClassesRepresentatives(G, N);
+    if Length(H)>0 then
+      if method in ["any", "str"] and not IsTrivial(N)
+                                  and not IsTrivial(H[1]) then
+        return [ N, H[1] ];
+      else
+        for i in [1..Length(H)] do
+          AddSet( NHs, [ N, H[i] ] );
+        od;
+      fi;
+    fi;
+  od;
+  if method in [ "any", "str" ] then
+    # no nontrivial decompositions exist
+    if fullNs then
+      SetSemidirectDecompositions(G,
+                      [ [TrivialSubgroup(G), G], [G, TrivialSubgroup(G)] ]);
+    fi;
+    return fail;
+  else
+    if fullNs then
+      SetSemidirectDecompositions(G, NHs);
+    fi;
+    return NHs;
+  fi;
+end);
 
```

Comment:
I have not reviwed this code in detail, but on a cursory glance it looks sane, and I trust that you know what you are doing here... :-)